File: app/src/main/java/com/psdk/CompileActivity.java
package com.psdk;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.ScrollView;
import android.widget.TextView;
import android.widget.Toast;

import java.io.IOException;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;

public class CompileActivity extends Activity {
    private PsdkProcessLauncher m_psdkProcessLauncher;
    private String m_applicationPath;
    private String m_internalWriteablePath;
    private String m_executionLocation;
    private String m_archiveLocation;
    private String m_outputArchiveLocation;

    private static final String SCRIPT = "compile.rb";
    private static final String CHECK_ENGINE_SCRIPT = "check_engine.rb";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.compiler);
        m_applicationPath = getApplicationInfo().dataDir;
        m_internalWriteablePath = getFilesDir().getPath();
        m_executionLocation = getIntent().getStringExtra("EXECUTION_LOCATION");
        m_archiveLocation = getIntent().getStringExtra("ARCHIVE_LOCATION");
        m_outputArchiveLocation = getIntent().getStringExtra("OUTPUT_ARCHIVE_LOCATION");

        TextView compilationLog = findViewById(R.id.compilationLog);
        ScrollView compilationScrollView = findViewById(R.id.compilationScrollView);
        compilationLog.setSelected(true);
        TextView compilationEndState = findViewById(R.id.compilationEndState);

        final Activity self = this;
        m_psdkProcessLauncher = new PsdkProcessLauncher(m_applicationPath) {
            @Override
            protected void accept(String lineMessage) {
                runOnUiThread(() -> {
                    compilationLog.append(lineMessage);
                    compilationLog.append("\n");
                    compilationScrollView.fullScroll(View.FOCUS_DOWN);
                });
            }

            @Override
            protected void onLogError(Exception error) {
                Toast.makeText(getApplicationContext(), error.getLocalizedMessage(), Toast.LENGTH_LONG).show();
            }
        };

        CompletionTask onCompleteCompilation = returnCode -> {
            if (returnCode == 0) {
                compilationEndState.setText("Compilation success !");

                try {
                    ZipUtility.zip(m_executionLocation + "/Release", m_outputArchiveLocation);
                    //removeRecursivelyDirectory(m_executionLocation + "/Release");

                    final Intent mainIntent = new Intent(self, MainActivity.class);
                    startActivity(mainIntent);
                } catch (Exception e) {
                    compilationEndState.setText("Unable to build the final archive: " + e.getLocalizedMessage());
                    return;
                }

            } else {
                compilationEndState.setText("Compilation failure");
            }
        };

        CompletionTask onCompleteCheckEngine = returnCode -> {
            if (returnCode != 0) {
                compilationEndState.setText("Check engine failure");
            }
        };

        try {
            //m_psdkProcessLauncher.runAsync(new PsdkProcess(this, CHECK_ENGINE_SCRIPT), buildPsdkProcessData(), onCompleteCheckEngine);
            m_psdkProcessLauncher.join();
            m_psdkProcessLauncher.runAsync(new PsdkProcess(this, SCRIPT), buildPsdkProcessData(), onCompleteCompilation);
        } catch (Exception ex) {
            Toast.makeText(getApplicationContext(), ex.getLocalizedMessage(), Toast.LENGTH_LONG).show();
        }
    }

    private void removeRecursivelyDirectory(String directoryPath) throws IOException {
        Path directory = Paths.get(directoryPath);
        Files.walkFileTree(directory, new SimpleFileVisitor<Path>() {
            @Override
            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
                Files.delete(file);
                return FileVisitResult.CONTINUE;
            }

            @Override
            public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
                Files.delete(dir);
                return FileVisitResult.CONTINUE;
            }
        });
    }

    @Override
    public void onBackPressed() {
        m_psdkProcessLauncher.killCurrentProcess();
    }

    private PsdkProcess.InputData buildPsdkProcessData() {
        return new PsdkProcess.InputData(m_internalWriteablePath, m_executionLocation, m_archiveLocation);
    }
}

File: app/src/main/java/com/psdk/CompileActivity.kt
package com.psdk

import android.app.Activity
import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.SharedPreferences
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.widget.Button
import android.widget.CheckedTextView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.result.contract.ActivityResultContracts
import com.psdk.zip.EpsaDecryptor
import java.util.zip.ZipFile
import java.io.File
import java.io.IOException

class CompileActivity: ComponentActivity() {

    private var m_archiveLocation: String? = null
    private var m_badArchiveLocation: String? = null
    private var m_releaseLocation: String? = null
    private var m_withSavedArchive: Boolean = false
    private var m_executionLocation: String? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.compiler)

        m_executionLocation = intent.getStringExtra("EXECUTION_LOCATION")
        m_releaseLocation = intent.getStringExtra("RELEASE_LOCATION")

        m_withSavedArchive = importedFile.canRead()
        setArchiveLocationValue(importedFile.path)

        val compileButton = findViewById<Button>(R.id.compileGame)
        compileButton.setOnClickListener { v: View? ->
            val compileIntent = Intent(this, CompileProcessActivity::class.java)
            compileIntent.putExtra("EXECUTION_LOCATION", m_executionLocation)
            compileIntent.putExtra("ARCHIVE_LOCATION", m_archiveLocation)
            compileGameActivityResultLauncher.launch(compileIntent)
            return@setOnClickListener
        }

        val locatePsdkButton = findViewById<View>(R.id.locatePSDK) as Button
        locatePsdkButton.setOnClickListener { v: View? -> selectProjectFile() }
    }

    private fun checkFilepathValid(filepath: String?): String? {
        if (filepath == null) {
            return "File path not provided"
        }
        var finalFile = File(filepath)
        if (!finalFile.exists() || !finalFile.canRead()) {
            return "Error : file at filepath $filepath does not exist or not readable"
        }

        // If this is an encrypted archive, decrypt it first
        if (EpsaDecryptor.isEpsaFile(finalFile)) {
            val decryptedFile = importedFile
            val error = EpsaDecryptor.decrypt(finalFile, decryptedFile)
            if (error != null) {
                return error
            }
            finalFile = decryptedFile
            m_archiveLocation = decryptedFile.path
        }

        val absPath = finalFile.absolutePath
        val sep = absPath.lastIndexOf(File.separator)
        val filename = absPath.substring(sep + 1).trim { it <= ' ' }
        return if (!filename.endsWith(".psa")) {
            "Error : selected file at filepath $filepath is not a 'psa' file : '$filename'"
        } else try {
            ZipFile(finalFile).use { zip ->
                if (zip.getEntry("pokemonsdk/version.txt") == null) {
                    "pokemonsdk folder not found in the archive"
                } else if (zip.getEntry("Game.rb") == null) {
                    "no Game.rb init script in the archive"
                } else null
            }
        } catch (e: IOException) {
            "Error while reading the archive: " + e.localizedMessage
        }
    }

    private fun setArchiveLocationValue(location: String?) {
        m_archiveLocation = location?.trim { it <= ' ' }

        val projectEngineHealth = findViewById<TextView>(R.id.projectEngineHealth)
        if (m_withSavedArchive) {
            lockScreenIfInvalidState(true)
            projectEngineHealth.text = "Archive is valid"
        } else {
            projectEngineHealth.text = "Processing archive..."
            lockScreenIfInvalidState(false)
            val thread: Thread = object : Thread() {
                override fun run() {
                    m_badArchiveLocation = checkFilepathValid(m_archiveLocation)
                    runOnUiThread {
                        val validState = lockScreenIfInvalidState(isValidState)
                        if (validState) {
                            projectEngineHealth.text = "Archive is valid"
                        } else if (m_badArchiveLocation != null) {
                            projectEngineHealth.text = m_badArchiveLocation
                        }
                    }
                }
            }
            thread.start()
        }
    }

    private fun lockScreenIfInvalidState(validState: Boolean): Boolean {
        val psdkLocationValid = findViewById<CheckedTextView>(R.id.psdkLocationValid)
        psdkLocationValid.isChecked = validState
        val clickButton = findViewById<Button>(R.id.compileGame)
        clickButton.isEnabled = validState
        return validState
    }

    private fun selectProjectFile() {
        try {
            chooseProjectFile()
        } catch (ex: ActivityNotFoundException) {
            Toast.makeText(applicationContext, "No suitable File Manager was found.", Toast.LENGTH_LONG).show()
        }
    }

    private fun chooseProjectFile() {
        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.type = "*/*"
        intent.addCategory(Intent.CATEGORY_OPENABLE)
        val chooseFile = Intent.createChooser(intent, "Choose a file")
        chooseProjectFileActivityResultLauncher.launch(chooseFile)
    }

    private val chooseProjectFileActivityResultLauncher =
        registerForActivityResult(
            ActivityResultContracts.StartActivityForResult()
        ) { result ->
            if (result.resultCode == Activity.RESULT_OK) {
                val uri = result.data?.data!!
                val stagingFile = importedStagingFile
                if (!stagingFile.exists()) {
                    stagingFile.createNewFile()
                }
                val out = stagingFile.outputStream()
                val input = contentResolver.openInputStream(uri)
                input?.copyTo(out)
                input?.close()
                out.close()
                m_withSavedArchive = false
                if (EpsaDecryptor.isEpsaFile(stagingFile)) {
                    // Encrypted archive: validation will decrypt to archive.psa
                    setArchiveLocationValue(stagingFile.path)
                } else {
                    // Plain archive: rename staging to final location
                    stagingFile.renameTo(importedFile)
                    setArchiveLocationValue(importedFile.path)
                }
            }
        }

    private val compileGameActivityResultLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) {}

    private val isValidState: Boolean
        get() = m_badArchiveLocation == null

    private val importedFile: File
        get() = File("$m_executionLocation/archive.psa")

    private val importedStagingFile: File
        get() = File("$m_executionLocation/archive.staging")

}